src/RedCode/Currency/Rate/Exception/BadXMLQueryException.php

```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace RedCode\Currency\Rate\Exception;
+
+use RedCode\Currency\Rate\Provider\ICurrencyRateProvider;
+
+class BadXMLQueryException extends BaseProviderException
+{
+    /**
+     * @var string
+     */
+    protected $query;
+
+    /**
+     * @param string $query
+     * @param ICurrencyRateProvider $provider
+     */
+    public function __construct($query, ICurrencyRateProvider $provider)
+    {
+        $this->query = $query;
+        $this->provider = $provider;
```

Comment:
It's better to hide all properties which are possible to hide.

So the solution using of base class constructor `parent::__construct($provider)` is more preferable.
This will also add strict dependency on `$provider` for base class.

In the base class change `protected $provider; ` to `private $provider;`
So the base class will be like this:

```PHP
abstract class BaseProviderException extends \Exception
{
    /**
     * @var \RedCode\Currency\Rate\Provider\ICurrencyRateProvider
     */
    private $provider;

    public function __construct(ICurrencyRateProvider $provider)
    {
        $this->provider = $provider;
    }

    /**
     * @return ICurrencyRateProvider
     */
    public function getProvider()
    {
        return $this->provider;
    }
}
```